src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1375 @@
+                OSMutexLock locker(&lock_);
+                if (mdnsClientCache_.size())
+                {
+                    // Trigger an immediate scan in order to refresh the cache.
+                    mdns_scanning_start_or_trigger_refresh();
+                }
+            }
+            char ip_addr[16];
+            inet_ntoa_r(d->ip_info.ip.addr, ip_addr, 16);
+            LOG(INFO, "wifi: STA got IP: %s", ip_addr);
+            ip_acquired(IFACE_STA, true);
+            break;
+        }
+        case IP_EVENT_STA_LOST_IP:
+        {
+            HASSERT(ipAcquiredSta_ == true);
+            ipAcquiredSta_ = false;
+            mdns_disable_sta();
+            LOG(INFO, "wifi: STA lost IP.");
+            ip_acquired(IFACE_STA, false);
+            break;
+        }
+        case IP_EVENT_AP_STAIPASSIGNED:
+            ipLeased_ = true;
+            break;
+    }
+}
+
+//
+// EspIdfWiFiBase::init_config()
+//
+void EspIdfWiFiBase::init_config()
+{
+    nvs_handle_t cfg;
+    esp_err_t result = nvs_open(NVS_NAMESPACE_NAME, NVS_READWRITE, &cfg);
+    if (result != ESP_OK)
+    {
+        LOG_ERROR("wifi: Error %s opening NVS handle.",
+            esp_err_to_name(result));
+        return;
+    }
+
+    // User configuration.
+    uint32_t magic;
+    size_t user_config_len;
+    void *user_config = get_user_config(&user_config_len, &magic);
+    size_t len = user_config_len;
+    result = nvs_get_blob(cfg, NVS_KEY_USER_NAME, user_config, &len);
+    switch(result)
+    {
+        case ESP_OK:
+            if (magic == WIFI_CONFIG_INIT_MAGIC && len == user_config_len)
+            {
+                // Already initialized.
+                break;
+            }
+            // Not initialized yet.
+            // fall through
+        case ESP_ERR_NVS_NOT_FOUND:
+            // fall through
+        case ESP_ERR_NVS_INVALID_LENGTH:
+            // Initialize WiFi user configuration to factory defaults.
+            factory_default();
+            nvs_set_blob(cfg, NVS_KEY_USER_NAME, user_config, user_config_len);
+            nvs_commit(cfg);
+            break;
+        default:
+            LOG_ERROR("wifi: Error %s getting userCfg_.",
+                esp_err_to_name(result));
+            break;
+    }
+
+    // Private configuration.
+    len = sizeof(privCfg_);
+    result = nvs_get_blob(cfg, NVS_KEY_LAST_NAME, &privCfg_, &len);
+    switch (result)
+    {
+        case ESP_OK:
+            if (privCfg_.magic_ == PRIV_CONFIG_INIT_MAGIC &&
+                len == sizeof(privCfg_))
+            {
+                // Already initialized.
+                break;
+            }
+            // Not initialized yet.
+            // fall through
+        case ESP_ERR_NVS_NOT_FOUND:
+            // fall through
+        case ESP_ERR_NVS_INVALID_LENGTH:
+            // Reset to factory defaults.
+            memset(&privCfg_, 0, sizeof(privCfg_));
+            nvs_set_blob(cfg, NVS_KEY_LAST_NAME, &privCfg_, sizeof(privCfg_));
```

Comment:
same here, i don't think we should write an empty blob.

Reply:
I've added a check such formatting the NVS is only allowed if a derived object says it is allowed, default implementation does not allow it. I've also updated this with the magic number, which was a bug.